src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java

```diff
@@ -0,0 +1,264 @@
+     * @param blockComponent          The block component of the vine entity
+     */
+    @ReceiveEvent
+    public void onPlantGrowth(OnVineGrowth event, EntityRef entity, VineDefinitionComponent vineDefinitionComponent, BlockComponent blockComponent) {
+        if (delayManager.hasDelayedAction(entity, GROW_ACTION)) {
+            delayManager.cancelDelayedAction(entity, GROW_ACTION);
+        }
+
+        if (!vineDefinitionComponent.isSapling) {
+            // Prefer repairing vines if broken
+            List<Vector3i> copy = new ArrayList<>(vineDefinitionComponent.parts);
+            for (Vector3i position: copy) {
+                if (worldProvider.getBlock(position) != blockManager.getBlock(vineDefinitionComponent.trunk)) {
+                    if (worldProvider.getBlock(position) == blockManager.getBlock(BlockManager.AIR_ID)) {
+                        Block trunkBlock = blockManager.getBlock(vineDefinitionComponent.trunk);
+                        worldProvider.setBlock(position, trunkBlock);
+                        scheduleVineGrowth(entity, vineDefinitionComponent);
+                        return;
+                    } else {
+                        vineDefinitionComponent.parts.remove(position);
+                    }
+                }
+            }
+
+            Vector3i newPartPosition = getNewPartPosition(vineDefinitionComponent);
+
+            if (newPartPosition == null) {
+                return;
+            }
+
+            if (vineDefinitionComponent.growthsTillRipe <= 0) {
+                if (Math.random() > 0.7) {
+                    Block fruitBlock = blockManager.getBlock(vineDefinitionComponent.produce);
+                    worldProvider.setBlock(newPartPosition, fruitBlock);
+                    scheduleVineGrowth(entity, vineDefinitionComponent);
+                    return;
+                }
+            }
+            if (vineDefinitionComponent.parts.size() <= vineDefinitionComponent.maxParts) {
+                Block trunkBlock = blockManager.getBlock(vineDefinitionComponent.trunk);
+                worldProvider.setBlock(newPartPosition, trunkBlock);
+                vineDefinitionComponent.parts.add(newPartPosition);
+                vineDefinitionComponent.growthsTillRipe = vineDefinitionComponent.growthsTillRipe - 1;
+            }
+        } else {
+            Block trunkBlock = blockManager.getBlock(vineDefinitionComponent.trunk);
+            Vector3i pos = vineDefinitionComponent.parts.get(0);
+            worldProvider.setBlock(pos, trunkBlock);
+            vineDefinitionComponent.isSapling = false;
+            EntityRef newPartEntity = blockEntityRegistry.getBlockEntityAt(pos);
+            newPartEntity.addComponent(vineDefinitionComponent);
+        }
+
+        scheduleVineGrowth(entity, vineDefinitionComponent);
+    }
+
+    private Vector3i getNewPartPosition(VineDefinitionComponent vineDefinitionComponent) {
+        java.util.Random randomGenerator = new java.util.Random();
```

Comment:
Killer 😎 
It works but can take a bit of cleanup:
- the list iteration can be replaced with shuffle and for-each
- the search does almost the same for each direction
- the method does not need the component as parameter ([Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter))


I tried a rewrite, based on the directions, which also checks the soil type of the ground (can be opt-out by passing null). The copy of the direction list is not required as long as it is a field but i don't like to change the value of parameters :)
I think it would also make sense to cache `blockManager.getBlock(BlockManager.AIR_ID)` instead of doing the lookup each time, Idid not change it in my extracted method so far.

Call it like `Vector3i newPartPosition = searchNewPartPosition(vineDefinitionComponent.parts, HORIZONTAL_DIRECTIONS, vineDefinitionComponent.soilCategory);`
with the constant
`private static List<Direction> HORIZONTAL_DIRECTIONS = Lists.newArrayList(Direction.FORWARD, Direction.BACKWARD, Direction.LEFT, Direction.RIGHT);`

This would also leave some extensibility for upward growing vines if we ever extract the directions to the prefab.
(Why not have a southberry, that grows always southwards ;) ).


```
/**
     * Searches for new plant part positions based on the current plant parts.
     * New positions are searched relative from each part in all passed directions in random order.
     * A valid position for a new plant part must met the following requirements:
     * <ul>
     * <li> The position is (a) adjacent to one part or (b) adjacent & one block below and has air above it
     * <li> The block at the new position is air
     * <li> The blow below the new position is not air
     * <li> The block below the new position has the required soil category for the part
     * </ul>
     * @param parts Position of the current plant parts/blocks
     * @param searchDirections Directions to search for, relative to each part
     * @param requiredSoilCategory The soil category, required for the ground below each part. Can be null to accept all kind of soils.
     * @return The new position for a plant part on a valid position or null if no valid position was found.
     */
    private Vector3i searchNewPartPosition(List<Vector3i> parts, List<Direction> searchDirections, String requiredSoilCategory) {

        List<Vector3i> partsCopy = new ArrayList<>(parts);
        Collections.shuffle(partsCopy);
        List<Direction> directionsCopy = new ArrayList<>(searchDirections);
        Collections.shuffle(directionsCopy);

        for (Vector3i part : partsCopy) {
            for (Direction direction : directionsCopy) {
                Vector3i possiblePosition = part.add(direction.getVector3i());
                //we are only interested in positions where the blocks are air
                if (isAirBlock(possiblePosition)) {
                    Vector3i oneDown = new Vector3i(possiblePosition).subY(1);
                    if (!isAirBlock(oneDown)) {
                        //one down is no air, so check the soil type next
                        if (requiredSoilCategory == null || hasBlockSoilCategory(oneDown, requiredSoilCategory)) {
                            //air at the position and valid soil type one down -> grow in this direction
                            return possiblePosition;
                        }
                    } else {
                        //one block down is air
                        Vector3i twoDown = new Vector3i(possiblePosition).subY(2);
                        if (!isAirBlock(twoDown) && (requiredSoilCategory == null || hasBlockSoilCategory(twoDown, requiredSoilCategory))) {
                            //two down is no air and has the required soil category -> grow at the block one down
                            //(crawling down shallow hills)
                            return oneDown;
                        }
                    }
                }
            }
        }
        return null;
    }

    private boolean isAirBlock(Vector3i blockPosition) {
        return worldProvider.getBlock(blockPosition) == blockManager.getBlock(BlockManager.AIR_ID);
    }

    private boolean hasBlockSoilCategory(Vector3i blockPosition, String soilCategory) {
        return worldProvider.getBlock(blockPosition).getBlockFamily().hasCategory(soilCategory);
    }
```